src/coordinator/functions/temporal/base.go

```diff
@@ -0,0 +1,348 @@
+)
+
+var emptyOp = baseOp{}
+
+// baseOp stores required properties for logical operations
+type baseOp struct {
+	operatorType string
+	duration     time.Duration
+	processorFn  MakeProcessor
+}
+
+func newBaseOp(args []interface{}, operatorType string, processorFn MakeProcessor) (baseOp, error) {
+	if len(args) != 1 {
+		return emptyOp, fmt.Errorf("invalid number of args for %s: %d", operatorType, len(args))
+	}
+
+	duration, ok := args[0].(time.Duration)
+	if !ok {
+		return emptyOp, fmt.Errorf("unable to cast to scalar argument: %v", args[0])
+	}
+
+	return baseOp{
+		operatorType: operatorType,
+		processorFn:  processorFn,
+		duration:     duration,
+	}, nil
+}
+
+// OpType for the operator
+func (o baseOp) OpType() string {
+	return o.operatorType
+}
+
+// String representation
+func (o baseOp) String() string {
+	return fmt.Sprintf("type: %s, duration: %v", o.OpType(), o.duration)
+}
+
+// Node creates an execution node
+func (o baseOp) Node(controller *transform.Controller, opts transform.Options) transform.OpNode {
+	return &baseNode{
+		controller:    controller,
+		cache:         transform.NewTimeCache(),
+		op:            o,
+		processor:     o.processorFn(o, controller),
+		transformOpts: opts,
+	}
+}
+
+// baseNode is an execution node
+type baseNode struct {
+	op            baseOp
+	controller    *transform.Controller
+	cache         *transform.TimeCache
+	processor     Processor
+	mu            sync.Mutex
+	transformOpts transform.Options
+}
+
+// Process processes a block. The processing steps are as follows:
+// 1. Figure out the maximum blocks needed for the temporal function
+// 2. For the current block, figure out whether we have enough previous blocks which can help process it
+// 3. For the blocks after current block, figure out which can be processed right now
+// 4. Process all valid blocks from #3, #4 and mark them as processed
+// 5. Run a sweep face to free up blocks which are no longer needed to be cached
+func (c *baseNode) Process(ID parser.NodeID, b block.Block) error {
+	iter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	meta := iter.Meta()
+	bounds := meta.Bounds
+	blockDuration := bounds.Duration
+	// Figure out the maximum blocks needed for the temporal function
+	maxBlocks := int(math.Ceil(float64(c.op.duration) / float64(blockDuration)))
+
+	// Figure out the leftmost block and rightmost blocks
+	leftRangeStart := bounds.Previous(maxBlocks).Start
+	queryStartBounds := bounds.Nearest(c.transformOpts.TimeSpec.Start)
+
+	if leftRangeStart.Before(queryStartBounds.Start) {
+		leftRangeStart = queryStartBounds.Start
+	}
+
+	rightRangeStart := bounds.Next(maxBlocks).Start
+	queryEndBounds := bounds.Nearest(c.transformOpts.TimeSpec.End.Add(-1 * bounds.StepSize))
+
+	if rightRangeStart.After(queryEndBounds.Start) {
+		rightRangeStart = queryEndBounds.Start
+	}
+
+	// Process left side of the range
+	leftBlks, emptyLeftBlocks := c.processLeft(b, bounds, maxBlocks, leftRangeStart)
```

Comment:
nit: this would read better if the bool returned by `processLeft` is `true` iff required blocks are present, rather than the inverse

Reply:
simplified some logic now, let me know how it reads